### PR TITLE
bicgstab: route through accumulator-aware mult()/dot()

### PR DESCRIPTION
**Pull request by Gurleen-kansray**
## Summary

This updates bicgstab.hpp to route its matrix-vector products and innerproducts through the accumulator-aware interfaces (mtl::mult<Accumulator>() and mtl::dot<Accumulator, value_type>()) instead of the plain A * x / naive dot() path, following the same pattern already applied in cg.hpp.

### Changes

- Added typename Accumulator = void to the template parameter list
- Replaced auto Ax = A * x; (+ copy loop) with a pre-allocated output vector passed to mtl::mult<Accumulator>(A, x, Ax)
- Replaced auto Aphat = A * phat; (+ copy loop) with mtl::mult<Accumulator>(A, phat, v)
- Replaced auto Ashat = A * shat; (+ copy loop) with mtl::mult<Accumulator>(A, shat, t)
- Routed all mtl::dot(...) calls through mtl::dot<Accumulator, value_type>(...)
- Added #include <mtl/operation/mult.hpp>

Default behavior (Accumulator = void) is unchanged.

### New test file

Added tests/unit/itl/test_bicgstab_quire.cpp, mirroring test_cg_quire.cpp:

- Sanity check: default Accumulator matches the unmodified baseline
- Edge case: 1x1 system
- Quire-vs-naive-posit32 accuracy comparison (compiled only whenMTL5_HAS_UNIVERSAL is defined)

### Verification

- itl_test_bicgstab: 17 assertions, 3 test cases -- all pass
- itl_test_bicgstab_quire (new): 6 assertions, 2 test cases -- all pass
- regression_itl_test_bicgstab: 12 assertions, 3 test cases, 6 scenarios
  (Laplacian-2D n=2500/10000, ConvDiff-1D n=10000/50000 with ILU(0),
  Laplacian-2D n=10000/24964 with Jacobi) -- all pass

### Related

Part of #254

## Summary by CodeRabbit

- **New Features**
  - Added configurable accumulation support to BiCGSTAB and GMRES for improved numerical flexibility.
  - Added clearer solver failure reporting for detected breakdown conditions.

- **Bug Fixes**
  - Fixed a GMRES edge case that could cause the solver to stall during exact breakdown scenarios.

- **Tests**
  - Added regression coverage for GMRES breakdown handling.
  - Adjusted BiCGSTAB test sizes and timeout settings to improve test reliability.

### CODE REVIEW

File: tests/unit/itl/test_bicgstab_quire.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        if (i > 0)     A(i,i-1) = Posit(-1.0);
+        if (i < n - 1) A(i,i+1) = Posit(-1.0);
+    }
+
+    vec::dense_vector<Posit> b(n, Posit(1.0));
+
+    vec::dense_vector<Posit> x_naive(n, Posit(0.0));
+    itl::pc::identity<mat::dense2D<Posit>> pc(A);
+    itl::basic_iteration<Posit> iter_naive(b, 200, Posit(1e-6));
+    itl::bicgstab(A, x_naive, b, pc, iter_naive); // default Accumulator = naive posit32
```
[the remark this answers is not in the document]
Comment (Gurleen-kansray) [reply tests/unit/itl/test_bicgstab_quire.cpp]: Fixed.....both bicgstab() calls now capture their return codes and assert `REQUIRE(err_naive == 0)` / `REQUIRE(err_quire == 0)` before the residual comparison, so a non-convergent solve fails fast and diagnosably instead of producing a flaky assertion.

File: tests/unit/itl/test_bicgstab_quire.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    itl::pc::identity<mat::dense2D<double>> pc(A);
+    itl::basic_iteration<double> iter(b, 100, 1e-10);
+
+    int err = itl::bicgstab(A, x, b, pc, iter);
+    REQUIRE(err == 0);
+    REQUIRE_THAT(x(0), Catch::Matchers::WithinAbs(0.5, 1e-8));
+}
+
+#ifdef MTL5_HAS_UNIVERSAL
```
Comment (Ravenwater) [inline tests/unit/itl/test_bicgstab_quire.cpp]: Oh dang, I wasn't paying attention. MTL5 cannot pull in Universal.

This was the core reason why we created mp-spice. that repo pulls in a clean Universal and a clean MTL5 and then does the quire bit.

We need to back this out.

**Review by Ravenwater — CHANGES_REQUESTED** [review]
We need to back Universal out of MTL5, create an mp-itl repo and place all the quire testing in that mp-itl repo.

File: tests/unit/itl/test_bicgstab_quire.cpp (hunk 2 above)
Comment (Ravenwater) [reply tests/unit/itl/test_bicgstab_quire.cpp]: Better idea: we still back out Universal out of MTL5, but we keep the Accumulator test in but use native float/double as our mixed-precision test in MTL5. That at least exercizes the Accumulator but removes the Universal dependency. 

I have created an mp-iterative repo where we can move the tests that pull in Universal posit and quire functionality.

**Review by Ravenwater — CHANGES_REQUESTED** [review]
leave the test, but test with float/double instead of posit/quire.

still remove Universal dependency. move posit/quire test to mp-iterative

File: tests/unit/itl/test_bicgstab_quire.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+#include <universal/number/posit/posit.hpp>
+#include <mtl/math/quire_accumulator.hpp>
```
Comment (Ravenwater) [inline tests/unit/itl/test_bicgstab_quire.cpp]: We need to write the accumulator_traits<> in such a way that it supports a regular accumulator/product, an FMA, and a quire.  That is the requirement for the traits in MTL.

Then they get used by the mp-iterative environment to test those possible dot product implementations.

**Review by Ravenwater — CHANGES_REQUESTED** [review]
need to remove quire_accumulator.hpp. This needs to be 'generalized' by the accumulator_traits alone. The types that the accumulator_traits accepts should cover all our requirements, which are product/accumulator, FMA, and quire models.

**Comment by Ravenwater** [comment]
@Gurleen-kansray I moved the test_cg_quire.cpp from this repo to the https://github.com/stillwater-sc/mp-iterative repo. That would be a pattern that for the moment we can repeat with bicgstab.

The architecture work we need to do is to write the accumulator_traits in such a way that we can support float/double, double/long double, int8/int32, etc., but also add FMA and quire accumulators. 

Looking at the quire API, that is currently not feasible as it has a specific convert_to() instead of a static_cast<Value> semantic. We need to figure out how we can generalize the accumulator_traits. That is going to be a pretty significant and important design.

Once we have that design, than MTL5 stays general, and we can refine mp-iterative to do the experiments. We will also remove the quire_accumulator.hpp specialization on which we currently lean.

File: tests/unit/itl/test_bicgstab_quire.cpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        return std::sqrt(res2);
+    };
+
+    double naive_residual = residual_norm(x_naive);
+    double wide_residual  = residual_norm(x_wide);
+
+    INFO("naive float32 residual: " << naive_residual);
+    INFO("float64-accumulated float32 residual: " << wide_residual);
+
+    REQUIRE(wide_residual <= naive_residual * 1.5);
```
[the remark this answers is not in the document]
Comment (Gurleen-kansray) [reply tests/unit/itl/test_bicgstab_quire.cpp]: This test file has been removed from MTL5 entirely (matching Ravenwater's approach for CG in #258) — the posit32+quire version now lives in mp-iterative instead. This comment no longer applies to the current diff.

File: tests/unit/itl/test_bicgstab_quire.cpp (hunk 2 above)
Comment (Gurleen-kansray) [reply tests/unit/itl/test_bicgstab_quire.cpp]: Done.. test_bicgstab_quire.cpp removed entirely from MTL5 (same as test_cg_quire.cpp), posit/quire coupling now lives in mp-iterative.

File: tests/unit/itl/test_bicgstab_quire.cpp (hunk 3 above)
Comment (Gurleen-kansray) [reply tests/unit/itl/test_bicgstab_quire.cpp]: Done — mirrored the same pattern here: test_bicgstab_quire.cpp removed from MTL5, posit32+quire version added to mp-iterative (stillwater-sc/mp-iterative, PR open).

File: tests/unit/itl/test_cg_quire.cpp [hunk 5]
```diff
@@ -52,49 +52,34 @@ TEST_CASE("CG handles 1x1 system", "[itl][cg][quire]") {
...
-    mat::dense2D<double> Ad(n, n);
-    vec::dense_vector<double> bd(n, 1.0);
-    for (std::size_t i = 0; i < n; ++i)
-        for (std::size_t j = 0; j < n; ++j)
-            Ad(i,j) = double(A(i,j));
+    vec::dense_vector<float> x_wide(n, 0.0f);
+    itl::basic_iteration<float> iter_wide(b, 200, 1e-4f);
+    itl::cg<mat::dense2D<float>, vec::dense_vector<float>, vec::dense_vector<float>,
+            itl::pc::identity<mat::dense2D<float>>, itl::basic_iteration<float>, double>(
+        A, x_wide, b, pc, iter_wide);
```
[the remark this answers is not in the document]
Comment (Gurleen-kansray) [reply tests/unit/itl/test_cg_quire.cpp]: Feedback addressed...test_bicgstab_quire.cpp and quire_accumulator.hpp fully removed from MTL5, matching your #258 pattern for CG. Ready for re-review.

[the remark this answers is not in the document]
Comment (Gurleen-kansray) [reply tests/unit/itl/test_cg_quire.cpp]: Confirmed, thanks.....agreed this finding is resolved by file removal.

[the remark this answers is not in the document]
Comment (Gurleen-kansray) [reply tests/unit/itl/test_cg_quire.cpp]: Confirmed, thanks.... agreed this finding is resolved by the file removal.

File: include/mtl/itl/krylov/gmres.hpp [hunk 6]
```diff
@@ -89,6 +85,17 @@ int gmres_inner(const LinearOp& A, VecX& x, const VecB& b,
...
 
+        if (breakdown) {
+            // Lucky breakdown: exact solution lies in the Krylov subspace
+            // spanned by V[0..k]. Column k's rotation was just applied
+            // above, so it's valid to include in back-substitution --
+            // unlike the old code, which broke out before applying the
+            // rotation or incrementing iter, leaving x and iter unchanged
+            // and causing gmres()'s outer restart loop to spin forever.
+            ++k;
+            break;
```
[the remark this answers is not in the document]
Comment (Gurleen-kansray) [reply include/mtl/itl/krylov/gmres.hpp:96]: This commit (gmres.hpp changes) has been removed from this branch —GMRES accumulator work will be a separate PR. This comment no longer applies to the current diff.

**Review by Ravenwater — APPROVED** [review]
let's merge and progress

**Comment by Gurleen-kansray** [comment]
@Ravenwater Thanks for your time and review Theo... :)
